### PR TITLE
8333230: Improve code example in doclet package

### PR DESCRIPTION
**Pull request by nizarbenalla**
Can I please get a review for this simple change? It's an attempt to improve the code example in doclet package.
Here are the generated docs with the [new code example.](https://cr.openjdk.org/~nbenalla/GeneratedDocs/newDocletExample/api/jdk.javadoc/jdk/javadoc/doclet/package-summary.html#Example.java)

I tried to go for simplicity while showing some features of the standard doclet, running this Doclet on a small project gives the following ouput.

```
Note: org.example.Example -> src/main/java/org/example/Example.java CLASS
Note: org.example.SimpleDoclet -> src/main/java/org/example/SimpleDoclet.java CLASS

```

TIA

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8333230](https://bugs.openjdk.org/browse/JDK-8333230): Improve code example in doclet package (**Bug** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20619/head:pull/20619` \
`$ git checkout pull/20619`

Update a local copy of the PR: \
`$ git checkout pull/20619` \
`$ git pull https://git.openjdk.org/jdk.git pull/20619/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20619`

View PR using the GUI difftool: \
`$ git pr show -t 20619`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20619.diff">https://git.openjdk.org/jdk/pull/20619.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20619#issuecomment-2293867269)

### CODE REVIEW

**Review by hns — COMMENTED** [review]
I think the new example is an improvement in some aspects, but it doesn't significantly simplify the example code, and in some other aspects it seems to add things that don't appear essential to me. 

Maybe we should define more clearly what the problem with the existing example code is, so we could either simplify the code further, or improve it in any other way that needs improvement.

One question we might ask is: do we need the example to implement custom options? It's certainly an interesting and useful feature, but it also adds a lot to the size of the code.

File: src/jdk.javadoc/share/classes/jdk/javadoc/doclet/package-info.java [hunk 1]
```diff
@@ -312,20 +273,50 @@
...
+ *
+ *    @Override
+ *    public boolean run(DocletEnvironment environment) {
+ *        treeUtils = environment.getDocTrees();
+ *        new ShowFileObjects(reporter).show(environment.getSpecifiedElements());
+ *        return OK;
+ *    }
+ *
+ *    // A scanner to loop through specified elements
+ *    class ShowFileObjects {
```
[the remark this answers is not in the document]
Comment (nizarbenalla) [reply src/jdk.javadoc/share/classes/jdk/javadoc/doclet/package-info.java:289]: I think it was mostly a coding style decision, I wanted to put emphasis on using a "scanner"

[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/doclet/package-info.java:289]: Yeah, I think we need a discussion on what is useful to appear in an example doclet.  I think the old code (lines 253-258) is more relevant to doclet processing than accessing filenames, but maybe there's not much in it.

File: src/jdk.javadoc/share/classes/jdk/javadoc/doclet/package-info.java [hunk 2]
```diff
@@ -312,20 +273,50 @@
...
+ *        final Reporter out;
+ *
+ *        ShowFileObjects(Reporter out) {
+ *            this.out = out;
+ *        }
+ *
+ *        void show(Set<? extends Element> elements) {
+ *            for (Element element : elements) {
+ *                if (element instanceof TypeElement te) {
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/doclet/package-info.java:298]: It is generally discouraged to use `instanceof` on `Element`s -- use `getKind` and a cast instead.
(yes, `instanceof` works on the JDK implementation but may not work on other implementations)

File: src/jdk.javadoc/share/classes/jdk/javadoc/doclet/package-info.java [hunk 3]
```diff
@@ -312,20 +273,50 @@
...
+ *                    TreePath dct = treeUtils.getPath(te);
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/doclet/package-info.java:299]: These days, you can use `Elements.getFileObjectOf(Element)`

**Comment by nizarbenalla** [comment]
Closing this for now until we decide how to go about this, it may be better to have a `Hello World` example than have the code run on itself

**Pull request closed without merging**